pydantic/_internal/_generate_schema.py

```diff
@@ -913,28 +901,43 @@ def _std_types_schema(self, obj: Any) -> core_schema.CoreSchema | None:
             except KeyError:
                 continue
             return encoder(self, obj)
-        if dataclasses.is_dataclass(obj):
-            return self._dataclass_schema(obj)  # type: ignore
         return None
 
-    def _dataclass_schema(self, dataclass: type[StandardDataclass]) -> core_schema.CoreSchema:
+    def _dataclass_schema(
+        self, dataclass: type[StandardDataclass], origin: type[StandardDataclass] | None
+    ) -> core_schema.CoreSchema:
         """
         Generate schema for a dataclass.
         """
-        # FIXME we need a way to make sure kw_only info is propagated through to fields
```

Comment:
did we make sure kw only constraints are kept in subclasses etc.?

Reply:
If I understand how it works correctly, I believe so, as I now read the `kw_only`-ness off the cls.__dataclass_fields__ FieldInfo, which seems to be there for _everything_ (including init_vars).